feat: implement unreachable calendar notifications system

Vansh5632 opened the pull request:
## What does this PR do?

This PR implements a comprehensive Unreachable Calendar Notifications System that detects when connected calendars become unreachable and notifies users via email with clear explanations and recovery instructions, while maintaining Cal.com's safety-first approach against overbooking.

- Fixes #23540 (GitHub issue number)
- Fixes CAL-6364 (Linear issue number - should be visible at the bottom of the GitHub issue description)

## Updates since last revision

This PR was completed by addressing merge conflicts and CodeRabbit review feedback:

### Merge Conflicts Resolved
- Merged upstream/main into the feature branch, resolving conflicts in:
  - `apps/web/components/apps/CalendarListContainer.tsx`
  - `packages/emails/email-manager.ts`
  - `packages/prisma/schema.prisma`
  - `packages/trpc/server/routers/viewer/me/get.handler.ts`
  - `packages/trpc/server/routers/viewer/me/updateProfile.schema.ts`

### CodeRabbit Review Feedback Addressed
1. **HTML Escaping**: Added `esc()` method to `UnreachableCalendarEmail` class to escape dynamic values preventing XSS vulnerabilities
2. **Named Export**: Changed `UnreachableCalendarEmail` from default to named export
3. **Optimistic Updates**: Refactored to use tRPC's `onMutate` callback for proper optimistic updates with rollback
4. **Loading State**: Added `isLoading` from tRPC query to disable Switch during loading
5. **Accessibility**: Added `aria-label` to Switch component
6. **Import Fix**: Fixed `CredentialRepository` import path in `markCredentialAsUnreachable.ts`

## Image
<img width="969" height="293" alt="Screenshot from 2025-09-07 16-28-47" src="https://github.com/user-attachments/assets/5e74513a-e4b3-41ad-8bc5-17fc94ed17e9" />

## Video

https://github.com/user-attachments/assets/3ff8fdc0-e13c-4c48-8195-75c516fed412

## Human Review Checklist

- [ ] Verify merge conflict resolution preserved all intended functionality
- [ ] Check optimistic update pattern works correctly with rapid toggles
- [ ] Verify HTML escaping doesn't break email rendering
- [ ] Confirm CredentialRepository import path is correct

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox.
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

---

**Link to Devin run**: https://app.devin.ai/sessions/c46997ef286d4cbfa6d98dd863562fed
**Requested by**: unknown ()
**Original author**: @Vansh5632

Vansh5632 committed:
feat: implement unreachable calendar notifications system

Vansh5632 commented:
okay i am working some changes are there after which i will put the video

Vansh5632 committed:
feat: integrate calendar notifications into main calendars settings

kart1ka committed:
Merge branch 'main' into feature/unreachable-calendar-notifications

Vansh5632 committed:
implemented the suggestions of coderabbit

kart1ka commented:
> @kart1ka video Added

The attached video does not quite work for us. That video does not show the complete feature actually working.

pallava-joshi reviewed (changes requested):
can you resolve the merge conflicts. also we have some failing unit tests.

devin-ai-integration[bot] committed:
Merge upstream/main into feature/unreachable-calendar-notifications

Resolved merge conflicts in:
- apps/web/components/apps/CalendarListContainer.tsx
- packages/emails/email-manager.ts
- packages/prisma/schema.prisma
- packages/trpc/server/routers/viewer/me/get.handler.ts
- packages/trpc/server/routers/viewer/me/updateProfile.schema.ts

Co-Authored-By: unknown <>

devin-ai-integration[bot] committed:
Address CodeRabbit review feedback for unreachable calendar notifications

- Add HTML escaping to email template to prevent XSS
- Change UnreachableCalendarEmail from default to named export
- Update email-manager.ts import for named export
- Improve CalendarListContainer.tsx with onMutate for optimistic updates
- Use isLoading from trpc query for better UX
- Add aria-label to Switch component for accessibility
- Fix CredentialRepository import path in markCredentialAsUnreachable.ts

Co-authored-by: Vansh5632 <vansh5632@users.noreply.github.com>
Co-Authored-By: unknown <>

sahitya-chandra reviewed (commented):
Closing this due to inactivity, let me know if you want to reopen it and resume the work
